File: 424-longest-repeating-character-replacement/main.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function(s, k) {
    /* There's no need to calculate this hashmap inside the for loop. We just need to make it global(define it here) and update it
    in case the l pointer moves forward. If you define it inside of for loop which means allocating it on each iteration of loop,
    we could have get time limit exceeded.*/
    const count = {}
    let res = 0
    let l = 0

    for (let r = 0; r < s.length; r++) {
        if (!count[s[r]]) {
            count[s[r]] = 1
        } else {
            count[s[r]]++
        }

        let max = -Infinity
        Object.values(count).forEach(v => {
            if (v > max) max = v
        })

        while (((r - l + 1) - max) > k) {
            count[s[l]]--
            l++
        }

        res = Math.max(res, r - l + 1)
    }

    return res
};

File: 424-longest-repeating-character-replacement/main.go
package main

// T: O(n)
// M: O(26)
func characterReplacement(s string, k int) int {
	count := make(map[byte]int)
	res := 0
	l := 0
	maxF := 0

	for r := 0; r < len(s); r++ {
		count[s[r]] += 1
		maxF = max(maxF, count[s[r]])

		/* shrink the window until it becomes valid. Why having a for loop and moving l possibly multiple times instead of
		just an if block and moving l once? It maybe the case that by moving r one time forward, the whole window becomes invalid and
		we have to move l forward multiple times. Like when having `aaabaaaa` with `k = 0`. The `res` is `4`. Now when we hit b, we have
		to move l all the way to b. So we need a for loop with moving l multiple times not just a single time. */
		for r-l+1-maxF > k {
			count[s[l]] -= 1
			l += 1
		}

		res = max(res, r-l+1)
	}

	return res
}
